batchflow/models/tf/efficientnet.py

```diff
@@ -0,0 +1,318 @@
+""" Base class for scalable models and EfficientNet
+
+Mingxing Tan, Quoc V. Le "`EfficientNet: Rethinking Model Scaling for Convolutional Neural Networks
+<https://arxiv.org/abs/1905.11946>`_"
+
+"""
+import numpy as np
+import tensorflow as tf
+
+from . import TFModel, MobileNet_v2
+from .layers import conv_block
+
+
+class ScalableModel(TFModel):
+    """ Base class for scalable models.
+    Scaling can be done by depth and/or by width
+
+    **Configuration**
+
+    inputs : dict
+        dict with 'images' and 'labels' (see :meth:`.TFModel._make_inputs`)
+
+    initial_block : dict
+        parameters for the initial block
+        base :
+            Base block. default is :func:`~.layers.conv_block`
+        repeats : int
+            number of repeats of the block
+        scalable : bool
+            indicates whether the block can be scaled
+        other parameters :
+            see :func:`~.layers.conv_block` or base block documentation
+
+    body : dict
+        base : function
+            Base block. default is :func:`~.layers.conv_block`
+        blocks : list of dict
+            parameters for each block
+            repeats : int
+                number of repeats of the block
+            scalable : bool
+                indicates whether the block can be scaled
+            other parameters :
+                see :func:`~.layers.conv_block` or base block documentation
+
+    head : dict
+        parameters for head
+
+    common : dict
+        common parameters
+        width_factor :
+        depth_factor : float
+            scaling factors
+    """
+
+    @classmethod
+    def default_config(cls):
+        config = super().default_config()
+
+        config['body'].update(dict(base=None,
+                                   blocks=[dict(repeats=1, scalable=True, layout='cna cna', kernel_size=3, filters=8)]))
+        config['head'].update(dict(layout='Pf'))
+
+        config['common/width_factor'] = 1.0
+        config['common/depth_factor'] = 1.0
+
+        config['loss'] = 'ce'
+        return config
+
+    def build_config(self, names=None):
+        config = super().build_config(names)
+
+        if config.get('head/units') is None:
+            config['head/units'] = self.num_classes('targets')
+
+        return config
+
+    @classmethod
+    def round_filters(cls, filters, factor):
+        """ Round number of filters based on width multiplier. """
+        return int(filters * factor)
+
+    @classmethod
+    def round_repeats(cls, repeats, factor):
+        """ Round number of layers based on depth multiplier. """
+        return int(np.ceil(repeats * factor))
+
+    @classmethod
+    def initial_block(cls, inputs, name='initial_block', **kwargs):
```

Comment:
Not needed

Reply:
Actually, you can bilinearly `resize` the input image to desired resolution right here, with `resolution_factor`  or `resolution` involved.

Reply:
It is needed for EfficientNets which have initial_blocks and heads that are width-scalable and not depth-scalabe. I'll move it there